### PR TITLE
EVG-5931: Commit Queue merge command

### PR DESCRIPTION
**Pull request by ybrill**
Theory of operation: 

1. an mbox formatted patch (which includes commit metadata) is uploaded as an unfinalized patch. 
2. the patch ID is enqueued on the project's commit queue.

steps 1 & 2 are severable (with the `--pause` option) to allow uploading modules to the patch before enqueuing (which is part of an upcoming ticket, [EVG-5932](https://jira.mongodb.org/browse/EVG-5932)).

Open question: do we want to allow passing in extra args to `git format-patch` as we do for `git diff` when creating patches?

### CODE REVIEW

File: operations/commit_queue.go [hunk 1]
```diff
@@ -10,7 +11,13 @@ import (
 	"github.com/urfave/cli"
 )
 
-const itemFlagName = "item"
+const (
+	itemFlagName       = "item"
+	refFlagName        = "ref"
+	pauseFlagName      = "pause"
+	identifierFlagName = "identifier"
+	commitQueueAlias   = "__commit_queue"
```
Comment (ybrill) [inline operations/commit_queue.go]: with this alias the eventual version will be populated with the variants/tasks configured for the project's commit queue.

File: operations/commit_queue.go [hunk 2]
```diff
@@ -86,6 +94,59 @@ func deleteItem() cli.Command {
...
+			params := mergeParams{
+				projectID:   projectID,
+				ref:         ref,
+				id:          id,
+				pause:       pause,
+				description: description,
+				skipConfirm: skipConfirm,
+				large:       largeOK,
+			}
+			return params.mergeBranch(ctx, confPath)
```
Comment (ybrill) [inline operations/commit_queue.go]: all the logic is removed from the Action (looking at Sam...)

File: operations/commit_queue.go [hunk 3]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+	if err != nil {
+		return errors.Wrap(err, "problem accessing evergreen service")
+	}
+
+	client := conf.GetRestCommunicator(ctx)
+	if err != nil {
+		return errors.Wrap(err, "problem accessing evergreen service")
+	}
+
+	if p.id == "" {
```
Comment (ybrill) [inline operations/commit_queue.go]: no id is provided: need to create the patch.

File: operations/commit_queue.go [hunk 4]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+		diffData, err := getFeaturePatchInfo(projectRef.Branch, p.ref)
+		if err != nil {
+			return errors.Wrap(err, "can't generate patches")
+		}
+
+		if p.description == "" && !p.skipConfirm {
+			p.description = prompt("Enter a description for this patch (optional):")
+		}
+		if p.description == "" {
+			p.description, err = gitCmd("rev-parse", "--abbrev-ref", p.ref)
```
Comment (ybrill) [inline operations/commit_queue.go]: if no description is provided we use the branch name, which is not equal to ref in the case of HEAD.

File: operations/commit_queue.go [hunk 5]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+			return err
+		}
+		grip.Infof("Queue position is %d", position)
+	}
+
+	return nil
+}
+
+func getFeaturePatchInfo(projectBranch, ref string) (*localDiff, error) {
+	upstream := projectBranch + "@{upstream}"
```
Comment (ybrill) [inline operations/commit_queue.go]: assume a local branch exists with 
- the same name as the project branch, and
- tracking the project branch

This is reasonable considering we already make this assumption for patches.

File: rest/client/rest.go [hunk 6]
```diff
@@ -711,6 +712,39 @@ func (c *communicatorImpl) DeleteCommitQueueItem(ctx context.Context, projectID,
 	return nil
 }
 
+func (c *communicatorImpl) EnqueueItem(ctx context.Context, id string) (int, error) {
+	info := requestInfo{
```
Comment (ybrill) [inline rest/client/rest.go]: implementing the REST route is a later ticket [(EVG-5933)](https://jira.mongodb.org/browse/EVG-5933).
This won't actually work until it's implemented.

File: operations/commit_queue.go (hunk 2 above)
Comment (tychoish) [reply operations/commit_queue.go]: sweet! I think it'd be fine to populate mergeParams{} directly from calls to methods on the cli.Context

File: operations/commit_queue.go [hunk 7]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+		if err != nil {
+			return err
+		}
+		diffData, err := getFeaturePatchInfo(projectRef.Branch, p.ref)
+		if err != nil {
+			return errors.Wrap(err, "can't generate patches")
+		}
+
+		if p.description == "" && !p.skipConfirm {
+			p.description = prompt("Enter a description for this patch (optional):")
```
Comment (tychoish) [inline operations/commit_queue.go]: I think the prompts are kind of weird and legacy, can we just error? 

relatedly, does it make sense to separate out the "validate the options" from the base implementation

File: operations/commit_queue.go [hunk 8]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+	projectID   string
+	ref         string
+	id          string
+	pause       bool
+	description string
+	skipConfirm bool
+	large       bool
+}
+
+func (p *mergeParams) mergeBranch(ctx context.Context, confPath string) error {
```
Comment (tychoish) [inline operations/commit_queue.go]: can we split up this function a a bit.

File: operations/commit_queue.go [hunk 9]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+
+	if p.id == "" {
+		if p.projectID == "" {
+			p.projectID = conf.FindDefaultProject()
+		} else {
+			if conf.FindDefaultProject() == "" &&
+				!p.skipConfirm && confirm(fmt.Sprintf("Make %v your default project?", p.projectID), true) {
+				conf.SetDefaultProject(p.projectID)
+				if err := conf.Write(""); err != nil {
+					grip.Warningf("warning - failed to set default project: %v\n", err)
```
Comment (tychoish) [inline operations/commit_queue.go]: perhaps use the message.WrapError()

File: operations/commit_queue.go [hunk 10]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+	if err != nil {
+		return errors.Wrap(err, "problem accessing evergreen service")
+	}
+
+	if p.id == "" {
+		if p.projectID == "" {
+			p.projectID = conf.FindDefaultProject()
+		} else {
+			if conf.FindDefaultProject() == "" &&
+				!p.skipConfirm && confirm(fmt.Sprintf("Make %v your default project?", p.projectID), true) {
```
Comment (tychoish) [inline operations/commit_queue.go]: do we need to duplicate this default setting behavior here?

File: operations/commit_queue.go [hunk 11]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+
+	ac, _, err := conf.getLegacyClients()
+	if err != nil {
+		return errors.Wrap(err, "problem accessing evergreen service")
+	}
+
+	client := conf.GetRestCommunicator(ctx)
+	if err != nil {
+		return errors.Wrap(err, "problem accessing evergreen service")
+	}
```
Comment (tychoish) [inline operations/commit_queue.go]: I'd probably move these kinds of dependency setup operations out of this function. 

also ensure that you don't wrap errors with the same message.

File: operations/commit_queue.go [hunk 12]
```diff
@@ -117,3 +178,124 @@ func deleteCommitQueueItem(ctx context.Context, client client.Communicator, proj
...
+	large       bool
+}
+
+func (p *mergeParams) mergeBranch(ctx context.Context, confPath string) error {
+	conf, err := NewClientSettings(confPath)
+	if err != nil {
+		return errors.Wrap(err, "problem loading configuration")
+	}
+
+	ac, _, err := conf.getLegacyClients()
```
Comment (tychoish) [inline operations/commit_queue.go]: what functionality do we need from the legacy client that we can't get from the new client?

File: operations/commit_queue.go (hunk 1 above)
Comment (ybrill) [reply operations/commit_queue.go]: done. In globals.go

File: operations/commit_queue.go (hunk 2 above)
Comment (ybrill) [reply operations/commit_queue.go]: done.

File: operations/commit_queue.go (hunk 7 above)
Comment (ybrill) [reply operations/commit_queue.go]: done. 
Refactored validating the project id in patches and reused it.

File: operations/commit_queue.go (hunk 8 above)
Comment (ybrill) [reply operations/commit_queue.go]: done.

File: operations/commit_queue.go (hunk 9 above)
Comment (ybrill) [reply operations/commit_queue.go]: I'm not sure: the error ends up printed on the command line.

File: operations/commit_queue.go (hunk 10 above)
Comment (ybrill) [reply operations/commit_queue.go]: Refactored validating the project id in patches and reused it.

File: operations/commit_queue.go (hunk 11 above)
Comment (ybrill) [reply operations/commit_queue.go]: done.

File: operations/commit_queue.go (hunk 12 above)
Comment (ybrill) [reply operations/commit_queue.go]: we use the legacy client to: 

1. get the projectRef
2. upload the patch

These routes aren't implemented in the new client, as far as I can tell.

File: rest/client/rest.go [hunk 13]
```diff
@@ -711,6 +712,39 @@ func (c *communicatorImpl) DeleteCommitQueueItem(ctx context.Context, projectID,
...
+	if resp.StatusCode != http.StatusOK {
+		restErr := gimlet.ErrorResponse{}
+		if err = json.Unmarshal(bytes, &restErr); err != nil {
+			return 0, errors.Errorf("received an error but was unable to parse: %s", string(bytes))
+		}
+
+		return 0, restErr
+	}
+
+	position, err := strconv.Atoi(string(bytes))
```
Comment (tychoish) [inline rest/client/rest.go]: shouldn't the result be in json rather than a single integer?

File: operations/commit_queue.go (hunk 9 above)
Comment (tychoish) [reply operations/commit_queue.go]: I think that's an artifact of the plainLogger that you set up, if you undo that and then do the wrap thing, I bet it'll work.

Comment (ybrill) [reply operations/commit_queue.go]: done.

File: rest/client/rest.go (hunk 13 above)
Comment (ybrill) [reply rest/client/rest.go]: done.

**Pull request merged**